File: src/Core/StagingFinancialEvents/Schemas/IdentifierRelationship.py
from dataclasses import dataclass
from enum import Enum
from typing import Sequence, Union

from arrow import Arrow

from Core.StagingFinancialEvents.Schemas.FinancialIdentifier import (
    StagingFinancialIdentifier,
)


class StagingIdentifierChangeType(Enum):
    """How a staging financial identifier was superseded by another.

    RENAME: Ticker/name change only; no quantity change.
    SPLIT: 1 share becomes 2 or more under the new identifier.
    REVERSE_SPLIT: 2 or more shares become 1 under the new identifier.
    """

    RENAME = "RENAME"
    SPLIT = "SPLIT"
    REVERSE_SPLIT = "REVERSE_SPLIT"

    UNKNOWN = "UNKNOWN"


@dataclass
class StagingIdentifierRelationship:
    """Directed edge: FromIdentifier was superseded/changed to ToIdentifier (staging identifiers; e.g. RENAME)."""

    FromIdentifier: StagingFinancialIdentifier
    ToIdentifier: StagingFinancialIdentifier
    ChangeType: StagingIdentifierChangeType
    """When the change took effect (e.g. corporate action date). Optional; some countries need it for reporting."""
    EffectiveDate: Arrow | None = None


@dataclass
class StagingIdentifierRelationshipSplit:
    """SPLIT or REVERSE_SPLIT: same shape as base plus required quantity context for scaling (standalone to avoid dataclass default ordering)."""

    FromIdentifier: StagingFinancialIdentifier
    ToIdentifier: StagingFinancialIdentifier
    ChangeType: StagingIdentifierChangeType
    EffectiveDate: Arrow | None
    QuantityBefore: float
    QuantityAfter: float


@dataclass
class StagingIdentifierRelationshipPartial:
    """
    One side of an identifier relationship, with a correlation key so it can be merged
    with another partial (from another row or file) into a full StagingIdentifierRelationship.
    Exactly one of FromIdentifier or ToIdentifier is set.
    """

    FromIdentifier: StagingFinancialIdentifier | None
    ToIdentifier: StagingFinancialIdentifier | None
    CorrelationKey: str
    ChangeType: StagingIdentifierChangeType
    EffectiveDate: Arrow
    """Absolute quantity for this side (from-side = before, to-side = after). Used for SPLIT/REVERSE_SPLIT."""
    Quantity: float | None = None


@dataclass
class StagingIdentifierRelationshipPartialWithQuantity:
    """Partial that carries a required quantity for this side (used when merging into StagingIdentifierRelationshipSplit).
    Standalone dataclass to avoid overriding base Partial's Quantity type (float | None) with float.
    """

    FromIdentifier: StagingFinancialIdentifier | None
    ToIdentifier: StagingFinancialIdentifier | None
    CorrelationKey: str
    ChangeType: StagingIdentifierChangeType
    EffectiveDate: Arrow
    Quantity: float


StagingIdentifierRelationshipPartialAny = Union[
    StagingIdentifierRelationshipPartial,
    StagingIdentifierRelationshipPartialWithQuantity,
]
StagingIdentifierRelationshipAny = Union[StagingIdentifierRelationship, StagingIdentifierRelationshipSplit]


@dataclass
class StagingIdentifierRelationships:
    """Container for full and partial staging identifier relationships.

    Full Relationships are ready to use; PartialRelationships can be merged (by CorrelationKey)
    into full relationships in a later, broker-agnostic step (e.g. after combining multiple files).
    """

    Relationships: Sequence[StagingIdentifierRelationshipAny]
    PartialRelationships: Sequence[StagingIdentifierRelationshipPartialAny]

File: lib/Core/Schemas/Staging/IdentifierRelationship.ts
import type { DateTime } from "luxon";
import type { StagingFinancialIdentifier } from "@brrr/Core/Schemas/Staging/StagingFinancialIdentifier.ts";

export enum StagingIdentifierChangeType {
	RENAME = "RENAME",
	SPLIT = "SPLIT",
	REVERSE_SPLIT = "REVERSE_SPLIT",
	UNKNOWN = "UNKNOWN",
}

type StagingIdentifierRelationshipArgs = {
	fromIdentifier: StagingFinancialIdentifier;
	toIdentifier: StagingFinancialIdentifier;
	changeType: StagingIdentifierChangeType;
	effectiveDate: DateTime | null;
};

export class StagingIdentifierRelationship {
	public readonly fromIdentifier: StagingFinancialIdentifier;
	public readonly toIdentifier: StagingFinancialIdentifier;
	public readonly changeType: StagingIdentifierChangeType;
	public readonly effectiveDate: DateTime | null;

	constructor(args: StagingIdentifierRelationshipArgs) {
		this.fromIdentifier = args.fromIdentifier;
		this.toIdentifier = args.toIdentifier;
		this.changeType = args.changeType;
		this.effectiveDate = args.effectiveDate;
	}

	copy(overrides: Partial<ConstructorParameters<typeof StagingIdentifierRelationship>[0]>): StagingIdentifierRelationship {
		return new StagingIdentifierRelationship({
			fromIdentifier: this.fromIdentifier,
			toIdentifier: this.toIdentifier,
			changeType: this.changeType,
			effectiveDate: this.effectiveDate,
			...overrides,
		});
	}
}

export class StagingIdentifierRelationshipSplit {
	public readonly fromIdentifier: StagingFinancialIdentifier;
	public readonly toIdentifier: StagingFinancialIdentifier;
	public readonly changeType: StagingIdentifierChangeType;
	public readonly effectiveDate: DateTime | null;
	public readonly quantityBefore: number;
	public readonly quantityAfter: number;

	constructor(args: StagingIdentifierRelationshipArgs & { quantityBefore: number; quantityAfter: number }) {
		this.fromIdentifier = args.fromIdentifier;
		this.toIdentifier = args.toIdentifier;
		this.changeType = args.changeType;
		this.effectiveDate = args.effectiveDate;
		this.quantityBefore = args.quantityBefore;
		this.quantityAfter = args.quantityAfter;
	}

	copy(overrides: Partial<ConstructorParameters<typeof StagingIdentifierRelationshipSplit>[0]>): StagingIdentifierRelationshipSplit {
		return new StagingIdentifierRelationshipSplit({
			fromIdentifier: this.fromIdentifier,
			toIdentifier: this.toIdentifier,
			changeType: this.changeType,
			effectiveDate: this.effectiveDate,
			quantityBefore: this.quantityBefore,
			quantityAfter: this.quantityAfter,
			...overrides,
		});
	}
}

type StagingIdentifierRelationshipPartialArgs = {
	fromIdentifier: StagingFinancialIdentifier | null;
	toIdentifier: StagingFinancialIdentifier | null;
	correlationKey: string;
	changeType: StagingIdentifierChangeType;
	effectiveDate: DateTime;
	quantity?: number | null;
};

export class StagingIdentifierRelationshipPartial {
	public readonly fromIdentifier: StagingFinancialIdentifier | null;
	public readonly toIdentifier: StagingFinancialIdentifier | null;
	public readonly correlationKey: string;
	public readonly changeType: StagingIdentifierChangeType;
	public readonly effectiveDate: DateTime;
	public readonly quantity?: number | null;

	constructor(args: StagingIdentifierRelationshipPartialArgs) {
		this.fromIdentifier = args.fromIdentifier;
		this.toIdentifier = args.toIdentifier;
		this.correlationKey = args.correlationKey;
		this.changeType = args.changeType;
		this.effectiveDate = args.effectiveDate;
		this.quantity = args.quantity;
	}

	copy(overrides: Partial<ConstructorParameters<typeof StagingIdentifierRelationshipPartial>[0]>): StagingIdentifierRelationshipPartial {
		return new StagingIdentifierRelationshipPartial({
			fromIdentifier: this.fromIdentifier,
			toIdentifier: this.toIdentifier,
			correlationKey: this.correlationKey,
			changeType: this.changeType,
			effectiveDate: this.effectiveDate,
			quantity: this.quantity,
			...overrides,
		});
	}
}

export class StagingIdentifierRelationshipPartialWithQuantity {
	public readonly fromIdentifier: StagingFinancialIdentifier | null;
	public readonly toIdentifier: StagingFinancialIdentifier | null;
	public readonly correlationKey: string;
	public readonly changeType: StagingIdentifierChangeType;
	public readonly effectiveDate: DateTime;
	public readonly quantity: number;

	constructor(args: Omit<StagingIdentifierRelationshipPartialArgs, "quantity"> & { quantity: number }) {
		this.fromIdentifier = args.fromIdentifier;
		this.toIdentifier = args.toIdentifier;
		this.correlationKey = args.correlationKey;
		this.changeType = args.changeType;
		this.effectiveDate = args.effectiveDate;
		this.quantity = args.quantity;
	}

	copy(
		overrides: Partial<ConstructorParameters<typeof StagingIdentifierRelationshipPartialWithQuantity>[0]>,
	): StagingIdentifierRelationshipPartialWithQuantity {
		return new StagingIdentifierRelationshipPartialWithQuantity({
			fromIdentifier: this.fromIdentifier,
			toIdentifier: this.toIdentifier,
			correlationKey: this.correlationKey,
			changeType: this.changeType,
			effectiveDate: this.effectiveDate,
			quantity: this.quantity,
			...overrides,
		});
	}
}

export type StagingIdentifierRelationshipPartialAny =
	| StagingIdentifierRelationshipPartial
	| StagingIdentifierRelationshipPartialWithQuantity;

export type StagingIdentifierRelationshipAny =
	| StagingIdentifierRelationship
	| StagingIdentifierRelationshipSplit;

export class StagingIdentifierRelationships {
	public readonly relationships: StagingIdentifierRelationshipAny[];
	public readonly partialRelationships: StagingIdentifierRelationshipPartialAny[];

	constructor(args: {
		relationships: StagingIdentifierRelationshipAny[];
		partialRelationships: StagingIdentifierRelationshipPartialAny[];
	}) {
		this.relationships = args.relationships;
		this.partialRelationships = args.partialRelationships;
	}
}
